### ISSUE TITLE
WAR transformation fails

### ISSUE THREAD

**Issue by rjahn** [eclipse/transformer#219]
I tried to transform a complete war file (simple structure). The file contains javax.servlet API and I tried to transform it to run on Tomcat 10. The WAR file runs without any problems on Tomcat <= 9.

I built my own snapshot based on a git clone (main branch) and with Java 1.8.0_144.

The command:
`...\org.eclipse.transformer.cli\target>java -jar org.eclipse.transformer.cli-0.3.0-SNAPSHOT.jar c:\temp\Projects5.war`

The output:
```
Copyright (c) Contributors to the Eclipse Foundation
This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
which is available at https://www.apache.org/licenses/LICENSE-2.0.
SPDX-License-Identifier: (EPL-2.0 OR Apache-2.0)

org.eclipse.transformer.cli.JakartaTransformerCLI
  Version [ 0.3.0-SNAPSHOT ]

[main] INFO Transformer - Input     [ c:/Temp/Projects5.war ]
[main] INFO Transformer -           [ c:\Temp\Projects5.war ]
[main] INFO Transformer - Output    [ c:/Temp/output_Projects5.war ] (Generated from input)
[main] INFO Transformer -           [ c:\Temp\output_Projects5.war ]
[main] INFO Transformer - Skipping option [ RULES_SELECTIONS ]
[main] INFO Transformer - Internal [ RULES_RENAMES] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-renames.properties ]
[main] INFO Transformer - Internal [ RULES_VERSIONS] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-versions.properties ]
[main] INFO Transformer - Internal [ RULES_BUNDLES] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-bundles.properties ]
[main] INFO Transformer - Internal [ RULES_DIRECT] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-direct.properties ]
[main] INFO Transformer - Internal [ RULES_MASTER_TEXT] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-text-master.properties ]
[main] INFO Transformer - Internal [ RULES_PER_CLASS_CONSTANT] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-per-class-constant-master.properties ]
[main] INFO Transformer - Package renames are in use
[main] INFO Transformer - Package versions will be updated
[main] INFO Transformer - Bundle identities will be updated
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.TransactionSynchronizationRegistryService.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-transactionSynchronizationRegistry.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.UserTransactionService.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ application-client.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ TransactionJavaColonHelper.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.transaction.services.RemoteTransactionControllerService.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-transactionManager.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ web.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ com.ibm.ws.webcontainer.v40.dd.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-servletVersion.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ UTService.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-userTransaction.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ *.taglib.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ application.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ TransactionManager.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-transactionManager.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ faces-config.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ ejb-jar.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ web-fragment.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Internal [ Substitions matching [ permissions.xml ]] URL [ jar:file:/Z:/Users/rjahn/ROOT/Entwicklung/other/eclipse_transformer/org.eclipse.transformer.cli/target/libs/org.eclipse.transformer.jakarta-0.3.0-SNAPSHOT.jar!/org/eclipse/transformer/jakarta/jakarta-xml-dd.properties ]
[main] INFO Transformer - Text files will be updated
[main] INFO Transformer - Java direct string updates will not be performed
[main] INFO Transformer - Per class constant mapping files are not enabled
[main] INFO Transformer - All resources will be selected
[main] INFO Transformer - Java direct string updates will not be performed
[main] INFO Transformer - Per class constant mapping files are not enabled
[main] INFO Transformer - Action selected for input [ c:/Temp/Projects5.war ]: WAR Action
[main] ERROR Transformer - Failed to parse manifest [ META-INF/MANIFEST.MF ]
java.io.IOException: line too long
        at java.util.jar.Attributes.read(Unknown Source)
        at java.util.jar.Manifest.read(Unknown Source)
        at java.util.jar.Manifest.<init>(Unknown Source)
        at org.eclipse.transformer.action.impl.ManifestActionImpl.apply(ManifestActionImpl.java:113)
        at org.eclipse.transformer.action.impl.ActionImpl.basicApply(ActionImpl.java:482)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:462)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:268)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:256)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:558)
        at org.eclipse.transformer.Transformer.transform(Transformer.java:1255)
        at org.eclipse.transformer.Transformer.run(Transformer.java:181)
        at org.eclipse.transformer.cli.TransformerCLI.run(TransformerCLI.java:393)
        at org.eclipse.transformer.cli.TransformerCLI.runWith(TransformerCLI.java:58)
        at org.eclipse.transformer.cli.JakartaTransformerCLI.main(JakartaTransformerCLI.java:26)
[main] ERROR Transformer - Failed to parse raw class bytes [ org/atmosphere/cpr/AtmosphereFramework.class ]
java.io.IOException: Not a valid class file (no CAFEBABE header)
        at aQute.bnd.classfile.ClassFile.parseClassFile(ClassFile.java:69)
        at org.eclipse.transformer.action.impl.ClassActionImpl.apply(ClassActionImpl.java:284)
        at org.eclipse.transformer.action.impl.ActionImpl.basicApply(ActionImpl.java:482)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:462)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:268)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:256)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:558)
        at org.eclipse.transformer.Transformer.transform(Transformer.java:1255)
        at org.eclipse.transformer.Transformer.run(Transformer.java:181)
        at org.eclipse.transformer.cli.TransformerCLI.run(TransformerCLI.java:393)
        at org.eclipse.transformer.cli.TransformerCLI.runWith(TransformerCLI.java:58)
        at org.eclipse.transformer.cli.JakartaTransformerCLI.main(JakartaTransformerCLI.java:26)
[main] ERROR Transformer - Unexpected failure:
java.lang.NullPointerException: name
        at java.util.Objects.requireNonNull(Unknown Source)
        at java.util.zip.ZipEntry.<init>(Unknown Source)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:275)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:256)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:558)
        at org.eclipse.transformer.Transformer.transform(Transformer.java:1255)
        at org.eclipse.transformer.Transformer.run(Transformer.java:181)
        at org.eclipse.transformer.cli.TransformerCLI.run(TransformerCLI.java:393)
        at org.eclipse.transformer.cli.TransformerCLI.runWith(TransformerCLI.java:58)
        at org.eclipse.transformer.cli.JakartaTransformerCLI.main(JakartaTransformerCLI.java:26)
Return Code [ 3 ]: Transform Error
```

There are different errors. The war doesn't contain a META-INF directory. So this shouldn't be a problem. The class `org/atmosphere/cpr/AtmosphereFramework.class` looks like a correct class file?

**Comment by rjahn** [comment]
The WAR file can be loaded from [here](https://ww2.sibvisions.com/files/download/Projects5.war).

**Comment by rjahn** [comment]
with release 0.2.0 the problems don't occur.

**Comment by papegaaij** [comment]
This bug is a regression caused by this commit: 54c8f1192c01f1d21945b9de34da94bf0acf4e15

After this change, the transformation fails consistently for various jars, including Google Protobuf 3.19.3: https://repo1.maven.org/maven2/com/google/protobuf/protobuf-java/3.19.3/protobuf-java-3.19.3.jar

```
[main] INFO Transformer - Action selected for input [ protobuf-java-3.19.3.jar ]: Jar Action
[main] ERROR Transformer - Failed to parse raw class bytes [ com/google/protobuf/MessageSchema.class ]
java.io.IOException: Not a valid class file (no CAFEBABE header)
        at aQute.bnd.classfile.ClassFile.parseClassFile(ClassFile.java:69)
        at org.eclipse.transformer.action.impl.ClassActionImpl.apply(ClassActionImpl.java:284)
        at org.eclipse.transformer.action.impl.ActionImpl.basicApply(ActionImpl.java:482)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:462)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:268)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:558)
        at org.eclipse.transformer.Transformer.transform(Transformer.java:1255)
        at org.eclipse.transformer.Transformer.run(Transformer.java:181)
        at org.eclipse.transformer.cli.TransformerCLI.run(TransformerCLI.java:393)
        at org.eclipse.transformer.cli.TransformerCLI.runWith(TransformerCLI.java:58)
        at org.eclipse.transformer.cli.JakartaTransformerCLI.main(JakartaTransformerCLI.java:26)
[main] ERROR Transformer - Unexpected failure:
java.lang.NullPointerException: name
        at java.base/java.util.Objects.requireNonNull(Objects.java:233)
        at java.base/java.util.zip.ZipEntry.<init>(ZipEntry.java:106)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:275)
        at org.eclipse.transformer.action.impl.ContainerActionImpl.apply(ContainerActionImpl.java:152)
        at org.eclipse.transformer.action.impl.ActionImpl.apply(ActionImpl.java:558)
        at org.eclipse.transformer.Transformer.transform(Transformer.java:1255)
        at org.eclipse.transformer.Transformer.run(Transformer.java:181)
        at org.eclipse.transformer.cli.TransformerCLI.run(TransformerCLI.java:393)
        at org.eclipse.transformer.cli.TransformerCLI.runWith(TransformerCLI.java:58)
        at org.eclipse.transformer.cli.JakartaTransformerCLI.main(JakartaTransformerCLI.java:26)
Return Code [ 3 ]: Transform Error
```

**Comment by arjantijms** [comment]
@papegaaij 
> This bug is a regression caused by this commit: https://github.com/eclipse/transformer/commit/54c8f1192c01f1d21945b9de34da94bf0acf4e15

Can confirm that. Pretty much everything I try to transform after that commit fails. The rev just before it works indeed.

@bjhargrave can you take a look? It would be awesome if this could be fixed, and at the same time a new 0.3.0 release could be done. It's been a while since 0.2.0 ;)
